Backend setup

PrathamS1 opened the pull request:
RealReview Backend – Setup & Deployment (Branch: `backend-setup`)

This is the backend service for **RealReview**, a real estate platform that allows users to upload property images with metadata such as location, timestamp, user, and rating. This branch focuses on backend setup, PostgreSQL integration, and server deployment to AWS EC2.

Features Implemented

- Upload an image with metadata (`location`, `submitted_by`, `timestamp`, `rating`)
- View all uploaded images along with metadata
- Delete an uploaded image
- Store image files on server and metadata in PostgreSQL
- API deployed and running on AWS EC2 (HTTP, port 80)

PrathamS1 committed:
basic crud functionality implemented in backend

PrathamS1 committed:
modified gitignore

PrathamS1 committed:
minor changes to crud operations

PrathamS1 committed:
Update README.md

cpandey05 commented on backend/controllers/image.js:
```diff
@@ -0,0 +1,66 @@
+        return res.status(400).json({ error: 'No Image is uploaded' });
+    }
+    try{
+        const result = await pool.query(
```
Why direct query in controller ? why not layers of service and repo?

PrathamS1 committed:
Refactored image controller by moving logic to separate service and repository layers

PrathamS1 committed:
Merge branch 'backend-setup' of https://github.com/PrathamS1/RealReview into backend-setup

PrathamS1 committed:
Implemented class based image model for db insertion

PrathamS1 committed:
removed commented codes

PrathamS1 commented:
Refactored the service and repo layer to use class-based logic for insertion, as suggested. Created the Image model, which now encapsulates all image-related data, and the service and repository layers have been updated accordingly.

PrathamS1 committed:
optimized queries to return selected columns

PrathamS1 committed:
Implemented the add ratings and view ratings feature to the Real Review backend along with a centralized error handling.

PrathamS1 committed:
added postman collection for testing the APIs to resource folder

PrathamS1 commented:
# RealReview Backend - Ratings Feature & Error Handling System

## 1. Ratings Feature Implementation

### Database Changes
- Created new `ratings` table with schema:
  - `id` (Primary Key)
  - `image_id` (Foreign Key referencing images table)
  - `rating` (Integer)
  - `timestamp` (Default current timestamp)

### Model Layer
- Implemented `Rating` model class with:
  - Properties: id, imageId, rating, timestamp
  - Constructor for creating rating instances
  - Methods for data validation

### Repository Layer
- Created `ratingRepo.js` with functions:
  - `insertRating`: Handles rating insertion and updates image's rating with the average rating by using a transaction to ensure data integrity and to make sure the changes to rating field in images database will reflect the addition of a new rating for that specific image.
  - `getRatingsByImageId`: Retrieves all ratings for a specific image
  - `getAverageRatingByImageId`: Calculates the average rating for a specific image
  - Implemented transaction support for atomic operations

### Service Layer
- Implemented `ratingService.js` with:
  - `addRating`: Validates and adds new ratings
  - `getImageRatings`: Retrieves ratings with average calculation
  - Business logic for rating validation
  - Error handling for invalid ratings

### Controller Layer
- Created `ratingController.js` with:
  - `rateImage`: Controller for rating an image
  - `getRating`: Controller for fetching ratings of an image

### Routes layer
  - POST `/api/images/:id/rate`: Add new rating
  - GET `/api/images/:id/ratings`: Get all ratings for an image
  - Proper request validation and response formatting

## 2. Centralized Error Handling System

### Error Types Definition
- Created `errors/errorTypes.js` with categorized errors:
  - Database Errors (e.g., INVALID_ID_FORMAT, UNIQUE_VIOLATION)
  - Image Errors (e.g., NOT_FOUND, UPLOAD_FAILED)
  - Rating Errors (e.g., INVALID_RATING, ADD_FAILED)
  - Generic Errors (e.g., SERVER_ERROR)

### Error Handler Implementation
- Created `errors/errorHandler.js` with:
  - `AppError` class extending Error
  - Specialized handlers for different error types:
    - `handleDatabaseError`: PostgreSQL error handling
    - `handleImageError`: Image-specific error handling
    - `handleRatingError`: Rating-specific error handling

### Service Layer Updates
- Updated `imageService.js` and `ratingService.js` to:
  - Use new error handling system
  - Implement consistent error patterns
  - Proper error transformation
  - Better error messages

### Controller Layer Updates
- Updated `imageController.js` and `ratingController.js` to:
  - Use centralized error handling
  - Standardize error responses
  - Proper error status codes
  - Consistent error message format

## 3. Postman APIs Collection
- `Resources` Folder - Added the Postman collection of APIs for testing the backend APIs where the backend has successfully been deployed on an EC2 instance in AWS.

PrathamS1 commented:
> Where have you hosted it? Did you use EC2? can you share postman and URL

Sir, I have hosted the backend on EC2 and the postman collection is included in the resources folder in the backend-setup branch.

i'm attaching files here also:
[Real Review APIs for Image and Rating.postman_collection.json](https://github.com/user-attachments/files/20422212/Real.Review.APIs.for.Image.and.Rating.postman_collection.json)
[Real Review Local.postman_environment.json](https://github.com/user-attachments/files/20422213/Real.Review.Local.postman_environment.json)

PrathamS1 committed:
Modified the backend to use S3 bucket for handling image files instead of previous method of uploading to local directory.

cpandey05 commented on backend/middleware/s3Uploader.js:
```diff
@@ -0,0 +1,42 @@
+const AWS = require('aws-sdk');
+
```
Why the package is middleware, why not is service? every server code is middleware .. this is part of service

cpandey05 commented on backend/middleware/s3Uploader.js:
```diff
@@ -0,0 +1,42 @@
+const AWS = require('aws-sdk');
+
+AWS.config.update({ region: 'us-east-1' });
+
```
what if user wants to use bucket in another region?

cpandey05 commented on backend/middleware/s3Uploader.js:
```diff
@@ -0,0 +1,42 @@
+  try {
+    await s3.upload(params).promise();
+    return true; // Return true on successful upload
+  } catch (error) {
```
is ti needed to return true or false? will not a successful completion me marked if there are no errors

cpandey05 commented on backend/middleware/s3Uploader.js:
```diff
@@ -0,0 +1,42 @@
+  } catch (error) {
+    console.error('Error uploading to S3:', error);
+    return false; // Return false on upload failure
+  }
```
instead of returning true or false, is there a way to propagate error

cpandey05 commented on backend/repository/imageRepo.js:
```diff
@@ -0,0 +1,43 @@
+
+//* This function inserts image data into the database
+const insertImageData=async(image)=>{
+    const res=await pool.query(
```
can you try integrating some ORM, so that it is mapped to class -optional, but if you want to try out, go ahead

PrathamS1 replied:
Sir, the class mapping is being performed in service layer:
```
const getImages = async () => {
  try {
    const images = await getAllImageData();
    return images.map((image) => {
      const imageObj = new Image(
        image.filename,
        image.location,
        image.submitted_by,
        image.rating
      );
      imageObj.id = image.id;
      imageObj.timestamp = image.timestamp;
      return imageObj;
    });
  } catch (error) {
    console.error("Error fetching images from image repo:", error);
    if (error instanceof AppError) {
      throw error;
    }
    // database errors
    if (error.code) {
      throw handleDatabaseError(error);
    }
    // other errors
    throw handleImageError(error);
  }
};
```

Do i need to move it to repo instead?

PrathamS1 committed:
moved s3 uploader to service layer

PrathamS1 committed:
Handled:
- Bucket Not Exist Error by adding a checkBucketExist function
- Added new Error Types related to S3
- Now the bucket region and name are specified in env file
- Moved error handling from s3Uploader to respective service where the functions are called.

PrathamS1 committed:
modified  image streaming controller to move error handling to the respective service function

PrathamS1 committed:
resolved image stream time out error

PrathamS1 commented:
I haven't added any lifecycle rules on S3 bucket objects yet. I tried adding the rule of deleting images after a certain days, but that also requires changes to the code to track this and delete the database records so I disabled that rule for now.

cpandey05 commented:
Thats exactly what i want to understand how will you connect lifecycle events to your code

PrathamS1 committed:
Changed database to Amazon DynamoDB from PostgreSQL. DynamoDB initial table setup implemented

PrathamS1 committed:
Migrate to DynamoDB and implement S3 auto-deletion

- Migrated from PostgreSQL to DynamoDB with CRUD operations
- Added S3 lifecycle rules for 5-day auto-deletion
- Implemented Lambda functions for database updates
- Set up EventBridge triggers and schedules

PrathamS1 commented:
# Migrated to DynamoDB and implemented S3 lifecycle automation

Major changes:
- Migrated database from PostgreSQL to DynamoDB
  - Updated models and schemas for DynamoDB
  - Implemented CRUD operations with DynamoDB
  - Added proper error handling for DynamoDB operations

- Implemented S3 lifecycle automation
  - Added S3 lifecycle rules for auto-deletion after 5 days
  - Created Lambda functions for database updates
    - archiveImageHandler for S3 lifecycle events
    - failedEventsHandler for failed event processing
  - Set up EventBridge rules and schedules
  - Implemented proper error handling and retry mechanism

Technical details:
- Used Dynamoose for DynamoDB operations
- Added transaction support for related operations
- Configured EventBridge triggers and schedules

cpandey05 commented on backend/database/db.js:
```diff
@@ -0,0 +1,29 @@
+    try {
+        console.log("Attempting to ensure Image and Rating tables exist and are active...");
+        await Image.scan().limit(1).exec();
+        console.log("Image table checked/created successfully.");
```
What does scan mean and why limit?
